### PR TITLE
CompositeJK Part 3: SplitJK

### PR DESCRIPTION
**Pull request by davpoolechem**
## Description
The final step of the CompositeJK saga.

The previous PR in this chain, https://github.com/psi4/psi4/pull/2833, introduced the CompositeJK framework to Psi4, a class by which to enable arbitrary mixing and matching of different algorithms for constructing the J and K matrices separately, with front-end changes to match. The big issue with the current formulation of CompositeJK, is that every single separate J or K build algorithm in Psi4 (currently DF-DirJ, LinK, and COSX) is contained within CompositeJK itself. This runs a high risk of CompositeJK becoming monolithic, especially as more separate J/K build methods are added (e.g., CFMM).

This PR is meant to be the solution to that issue. Rather than storing every separate J or K build algorithm in CompositeJK itself, they are now implemented as derived classes of a new base class, SplitJK. SplitJK features the `build_G_component` member function, which derived classes of SplitJK then implement to execute their own algorithm for building one of J or K. CompositeJK is reworked to match, featuring a pair of shared_ptrs to SplitJK classes, one representing the J algorithm, and one representing the K algorithm. Actual machinery for computing J or K, as well as other functionalities such as header printing, is then done through calling corresponding functions from the SplitJK pointers.

My thought is for this PR to be added in v1.9.

## User API & Changelog headlines
N/A

## Dev notes & details
- [x] Implements the SplitJK class for representing algorithms that build one of J or K.
- [x] Refactors DF-DirJ, LinK, and COSX as derived classes of SplitJK.
- [x] Reworks the CompositeJK class to execute in terms of SplitJK derived classes.

## Questions
- [ ] Is passing the eri_computers variable to COSX::build_G_component() as an argument acceptable, even though it is unused in that function? My logic, for now, is that the SplitJK implementation allows for a universal interface for calling different SplitJK derived classes, and thus different composite algorithm combinations; but this is a case where specifically fine-tuning the input for different methods may be preferable. What does everyone think?

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [ ] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/libfock/COSK.cc [hunk 1]
```diff
@@ -0,0 +1,706 @@
...
+            }
+
+            esp_boundp[s1][s2] = std::abs(esp_boundp[s1][s2]);
+        }
+    }
+
+    return esp_bound;
+}
+
+COSK::COSK(std::shared_ptr<BasisSet> primary, Options& options) : SplitJK(primary, options) {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/COSK.cc:150]: Machinery pulled from the previous CompositeJK::common_init().

File: psi4/src/psi4/libfock/COSK.cc [hunk 2]
```diff
@@ -0,0 +1,706 @@
...
+    timer_off("COSK: Setup");
+}
+
+COSK::~COSK() {}
+
+size_t COSK::num_computed_shells() {
+    return num_computed_shells_;
+}
+
+void COSK::print_header() const {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/COSK.cc:295]: Machinery pulled from the previous CompositeJK::print_COSX_header().

File: psi4/src/psi4/libfock/COSK.cc [hunk 3]
```diff
@@ -0,0 +1,706 @@
...
+        outfile->Printf("    K Density Cutoff:   %11.0E\n", dscreen_);
+        outfile->Printf("    K Basis Cutoff:     %11.0E\n", basis_tol_);
+        outfile->Printf("    K Overlap Fitting:  %11s\n", (overlap_fitted_ ? "Yes" : "No"));
+    }
+}
+
+// build the K matrix using Neeses's Chain-of-Spheres Exchange algorithm
+// algorithm is originally proposed in https://doi.org/10.1016/j.chemphys.2008.10.036
+// overlap fitting is discussed in https://doi.org/10.1063/1.3646921
+void COSK::build_G_component(std::vector<std::shared_ptr<Matrix>>& D, std::vector<std::shared_ptr<Matrix>>& K,
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/COSK.cc:310]: Machinery pulled from the previous CompositeJK::build_COSK.

File: psi4/src/psi4/libfock/DirectDFJ.cc [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+#include <algorithm>
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+using namespace psi;
+
+namespace psi {
+
+DirectDFJ::DirectDFJ(std::shared_ptr<BasisSet> primary, std::shared_ptr<BasisSet> auxiliary, Options& options) : SplitJK(primary, options), auxiliary_(auxiliary) {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/DirectDFJ.cc:50]: Machinery pulled from the previous CompositeJK::common_init() function.

File: psi4/src/psi4/libfock/DirectDFJ.cc [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    timer_off("DirectDFJ: Setup");
+}
+
+DirectDFJ::~DirectDFJ() {}
+
+size_t DirectDFJ::num_computed_shells() {
+    return num_computed_shells_;
+}
+
+void DirectDFJ::print_header() const {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/DirectDFJ.cc:79]: Machinery pulled from the previous CompositeJK::print_DirectDFJ_header() function.

File: psi4/src/psi4/libfock/DirectDFJ.cc [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        outfile->Printf("\n");
+        outfile->Printf("  ==> DF-DirJ: Integral-Direct Density-Fitted J <==\n\n");
+
+        outfile->Printf("    J Screening Cutoff:%11.0E\n", cutoff_);
+    }
+}
+
+// build the J matrix using Weigend's integral-direct density fitting algorithm
+// algorithm is in Figure 1 of https://doi.org/10.1039/B204199P
+void DirectDFJ::build_G_component(std::vector<std::shared_ptr<Matrix>>& D, std::vector<std::shared_ptr<Matrix>>& J,
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/DirectDFJ.cc:90]: Machinery pulled from the previous CompositeJK::build_DirectDFJ() function.

File: psi4/src/psi4/libfock/LinK.cc [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+#include <algorithm>
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+using namespace psi;
+
+namespace psi {
+
+LinK::LinK(std::shared_ptr<BasisSet> primary, Options& options) : SplitJK(primary, options) {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/LinK.cc:56]: Machinery pulled from the previous CompositeJK::common_init() function.

File: psi4/src/psi4/libfock/LinK.cc [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+    timer_off("LinK: Setup");
+}
+
+LinK::~LinK() {}
+
+size_t LinK::num_computed_shells() {
+    return num_computed_shells_;
+}
+
+void LinK::print_header() const {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/LinK.cc:86]: Machinery pulled from the previous CompositeJK::print_linK_header() function, unmodified.

File: psi4/src/psi4/libfock/LinK.cc [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+    outfile->Printf("\n");
+    outfile->Printf("  ==> LinK: Linear Exchange K <==\n\n");
+
+    outfile->Printf("    K Screening Cutoff:%11.0E\n", linK_ints_cutoff_);
+}
+
+// build the K matrix using Ochsenfelds's Linear Exchange (LinK) algorithm
+// To follow this code, compare with figure 1 of DOI: 10.1063/1.476741
+void LinK::build_G_component(std::vector<std::shared_ptr<Matrix>>& D, std::vector<std::shared_ptr<Matrix>>& K,
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/LinK.cc:95]: Machinery pulled from the previous CompositeJK::build_linK() function.

File: psi4/src/psi4/libfock/SplitJK.cc [hunk 10]
```diff
@@ -0,0 +1,62 @@
+/*
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/SplitJK.cc:1]: Largely new additions in this file.

File: psi4/src/psi4/libfock/SplitJK.h [hunk 11]
```diff
@@ -0,0 +1,301 @@
+/*
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/SplitJK.h:1]: Largely new additions in this file.

File: psi4/src/psi4/libfock/jk.h [hunk 12]
```diff
@@ -1344,15 +1298,13 @@ class PSI_API CompositeJK : public JK {
...
 
     // => Knobs <= //
+    std::string name() override { return "CompositeJK"; }
+
     /**
     * Print header information regarding JK
     * type on output file
     */
     void print_header() const override;
-
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/jk.h:1352]: Moved to corresponding SplitJK derived classes, as print_header()

File: psi4/src/psi4/libfock/CompositeJK.cc [hunk 13]
```diff
@@ -51,100 +51,6 @@ using namespace psi;
 
 namespace psi {
 
-Matrix compute_numeric_overlap(const DFTGrid &grid, const std::shared_ptr<BasisSet> &primary) {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/CompositeJK.cc:54]: Moved to COSK.cc

File: psi4/src/psi4/libfock/CompositeJK.cc [hunk 14]
```diff
@@ -195,167 +101,45 @@ void CompositeJK::common_init() {
...
+        for(int rank = 1; rank < nthreads_; rank++) {
+            eri_computers_["3-Center"][rank] = std::shared_ptr<TwoBodyAOInt>(eri_computers_["3-Center"].front()->clone());
+        }
     } else {
         throw PSIEXCEPTION("Invalid Composite J algorithm selected!");
     }
 
     // => Set up separate K algorithm <= //
 
-    // Linear Exchange (LinK)
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/CompositeJK.cc:233]: Code block moved to libfock/LinK.cc.

File: psi4/src/psi4/libfock/CompositeJK.cc [hunk 15]
```diff
@@ -195,167 +101,45 @@ void CompositeJK::common_init() {
...
-    if (k_type_ == "LINK") {
-        // set up LinK integral tolerance
-        if (options_["LINK_INTS_TOLERANCE"].has_changed()) {
-            linK_ints_cutoff_ = options_.get_double("LINK_INTS_TOLERANCE");
-        } else {
-            linK_ints_cutoff_ = cutoff_;
-        }
-
-    // Chain-of-Spheres Exchange (COSX)
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/CompositeJK.cc:242]: Code block moved to libfock/COSK.cc.

File: psi4/src/psi4/libfock/CompositeJK.cc [hunk 16]
```diff
@@ -386,46 +170,15 @@ void CompositeJK::print_header() const {
...
         if (do_K_) {
-            if (k_type_ == "LINK") { print_linK_header(); }
-            else if (k_type_ == "COSX") { print_COSX_header(); }
+            k_algo_->print_header();
         }
         outfile->Printf("\n");
     }
 }
 
-void CompositeJK::print_DirectDFJ_header() const {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/CompositeJK.cc:399]: Function moved to `DirectDFJ::print_header()` in libfock/DirectDFJ.cc.

File: psi4/src/psi4/libfock/CompositeJK.cc [hunk 17]
```diff
@@ -386,46 +170,15 @@ void CompositeJK::print_header() const {
...
-    if (print_) {
-        outfile->Printf("\n");
-        outfile->Printf("  ==> DF-DirJ: Integral-Direct Density-Fitted J <==\n\n");
-
-        outfile->Printf("    J Screening Cutoff:%11.0E\n", cutoff_);
-    }
-}
-
-void CompositeJK::print_linK_header() const {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/CompositeJK.cc:408]: Function moved to `LinK::print_header()` in libfock/LinK.cc.

File: psi4/src/psi4/libfock/CompositeJK.cc [hunk 18]
```diff
@@ -386,46 +170,15 @@ void CompositeJK::print_header() const {
...
-    if (print_) {
-        outfile->Printf("\n");
-        outfile->Printf("  ==> LinK: Linear Exchange K <==\n\n");
-
-        outfile->Printf("    K Screening Cutoff:%11.0E\n", linK_ints_cutoff_);
-    }
-}
-
-void CompositeJK::print_COSX_header() const {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/CompositeJK.cc:417]: Function moved to `COSK::print_header()` in libfock/COSK.cc.

File: psi4/src/psi4/libfock/CompositeJK.cc [hunk 19]
```diff
@@ -532,1059 +291,4 @@ void CompositeJK::compute_JK() {
 
 void CompositeJK::postiterations() {}
 
-// build the J matrix using Weigend's integral-direct density fitting algorithm
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/CompositeJK.cc:535]: Function moved to `DirectDFJ::build_G_component()` function in libfock/DirectDFJ.cc.

File: psi4/src/psi4/libfock/CompositeJK.cc [hunk 20]
```diff
@@ -532,1059 +291,4 @@ void CompositeJK::compute_JK() {
...
-    for(size_t jki = 0; jki < njk; jki++) {
-        for (size_t thread = 0; thread < nthreads_; thread++) {
-            J[jki]->add(JT[jki][thread]);
-        }
-        J[jki]->hermitivitize();
-    }
-
-}
-
-// build the K matrix using Ochsenfelds's Linear Exchange (LinK) algorithm
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/CompositeJK.cc:776]: Function moved to `LinK::build_G_component()` function in libfock/LinK.cc.

File: psi4/src/psi4/libfock/CompositeJK.cc [hunk 21]
```diff
@@ -532,1059 +291,4 @@ void CompositeJK::compute_JK() {
...
-    if (get_bench()) {
-        computed_shells_per_iter_["Quartets"].push_back(num_computed_shells());
-    }
-}
-
-
-// build the K matrix using Neeses's Chain-of-Spheres Exchange algorithm
-// algorithm is originally proposed in https://doi.org/10.1016/j.chemphys.2008.10.036 
-// overlap fitting is discussed in https://doi.org/10.1063/1.3646921
-void CompositeJK::build_COSK(std::vector<std::shared_ptr<Matrix>>& D, std::vector<std::shared_ptr<Matrix>>& K) {
```
Comment (davpoolechem) [inline psi4/src/psi4/libfock/CompositeJK.cc:1186]: Function moved to `COSK::build_G_component()` function in libfock/COSK.cc.

**Comment by davpoolechem** [comment]
All right, I shall officially open this up for review.

**Review by loriab — APPROVED** [review]
code chunk migrations sound good, and that seems most of the PR, so lgtm. hopefully code has at last reached a long-term home. :-) The signature standardization you mention sounds ok to me.

**Review by jturney — APPROVED** [review]
Looks good!

**Pull request merged**